### PR TITLE
Update episode 5 and 6

### PR DESCRIPTION
**Pull request by ALuesink**
Still in progress

### CODE REVIEW

**Comment by ALuesink** [comment]
Episode 5 - Documentation and Citation is migrated to the intermediate tutorial (https://github.com/common-workflow-lab/cwl-intermediate-tutorial/pull/7)

**Comment by ALuesink** [comment]
I've added a new episode at the end called More information with links to useful websites

**Comment by mr-c** [comment]
@gcapes Thanks for all your assistance! Can you review this PR? Since it is bigger I want to finish it before reviewing yours

**Comment by gcapes** [comment]
Sure - I'll start on it tomorrow. FWIW my PRs focus on the earlier episodes so could be reviewed independently of this one.

File: _episodes/debug.md [hunk 1]
```diff
@@ -10,16 +10,64 @@ objectives:
...
+~~~
+cwltool --validate CWL_SCRIPT.cwl
+~~~
+{: .language-bash}
+
+It is possible that the script is validated, however, it still gets an error. 
+If you encounter an error, the best practice is to run the workflow with the `--debug` flag.
+This will provide you with extensive information on the error you encounter.
+~~~
+cwltool --debug` CWL_SCRIPT.cwl
```
Comment (gcapes) [inline _episodes/debug.md]: extraneous backtick after `--debug`

File: _episodes/debug.md [hunk 2]
```diff
@@ -10,16 +10,64 @@ objectives:
...
+> By the end of this episode, learners should be able to
+> __recognize and fix simple bugs in their workflow code__.
+{: .callout}
+
+When working on a CWL workflow, you will probably encounter errors. There are many different errors possible.
+It is always very important to check the error message in the terminal, because it will give you information on the error.
+This error message will give you the type of error as well as the line of code that contains the error.
+Some of these errors will be explained in this episode. 
+
+A first step to check if your CWL script contains any errors, you can run the workflow with the `--validate` flag.
```
Comment (gcapes) [inline _episodes/debug.md]: Improved wording suggestion: As a first step ...

File: _episodes/debug.md [hunk 3]
```diff
@@ -10,16 +10,64 @@ objectives:
...
+![]({{page.root}}/fig/YAML_error_tab.png)
+- Typos in field names. It is very easy to forget for example the capital letters in field names.
+Errors with typos in field names will show `invalid field`.
+![]({{page.root}}/fig/YAML_error_typo_fieldname.png)
+- Typos in variable names. Similar to typos in field names, it is easy to make a mistake in referencing to a variable.
+These errors will show `Field references unknown identifier.`
+![]({{page.root}}/fig/YAML_error_typo_variable.png)
+
+### Wiring error
+Wiring errors often occur when you forget to add an output from a workflows step to the `outputs` section.
```
Comment (gcapes) [inline _episodes/debug.md]: workflows -> workflow's

File: _episodes/debug.md [hunk 4]
```diff
@@ -10,16 +10,64 @@ objectives:
...
+Type errors take place when there is a mismatch in type between variables. 
+When you declare a variable in the `inputs` section, the type of this variable has to match the type in the YAML inputs file 
+and the type used in one of the workflows steps. 
+The error message that is shown when this error occurs will tell you on which line the mismatch happens.
+
+![]({{page.root}}/fig/Type_error.png)
+
+### Format error
+Some files need a specific format that needs to be specified in the YAML inputs file, for example the fastq file in the RNA-seq analysis.
+When you don't specify a format, an error will occur. You can for example use the [EDAM](https://www.ebi.ac.uk/ols/ontologies/edam)ontology) ontology.
```
Comment (gcapes) [inline _episodes/debug.md]: fix link formatting - should be:
`[EDAM](https://www.ebi.ac.uk/ols/ontologies/edam) ontology.`

**Review by gcapes — DISMISSED** [review]
This looks like a good start!
I've made a few specific wording and typo comments, but in addition would it be useful to have a section on array vs single-item mismatch? That was in the original outline.

I would also suggest runnable examples here. Perhaps as exercises, and/or the input files for the error outputs shown.

Finally, I think that code output boxes would be preferable to screenshots. Certainly for maintainability - if someone else wants to add/modify the example it is likely their screenshots will look different.

**Comment by gcapes** [comment]
I've now fixed all the suggestions, except I can't reproduce the error from "Using tabs instead of spaces".

I also seem unable to push to this PR's branch, so I'll make a new PR to replace this one.

**Comment by mr-c** [comment]
> I've now fixed all the suggestions, except I can't reproduce the error from "Using tabs instead of spaces".
> 
> I also seem unable to push to this PR's branch, so I'll make a new PR to replace this one.

Did you try `gh pr checkout 56` using https://cli.github.com/ ?

**Comment by gcapes** [comment]
No, I've not used the github cli, but that command would only check out the branch (wouldn't it?), which I've already done. I've locally rebased onto gh-pages, and made the fixes.

**Comment by mr-c** [comment]
> No, I've not used the github cli, but that command would only check out the branch (wouldn't it?), which I've already done. I've locally rebased onto gh-pages, and made the fixes.

It sets up the remote to push as well. Anyhow, next time

**Comment by gcapes** [comment]
Look like I just figured it out :) I'd renamed my local branch, and it looks like that has to match the remote branch name.

**Comment by gcapes** [comment]
Can someone suggest how to reproduce the tabs instead of spaces error?

**Comment by mr-c** [comment]
@ALuesink do you still have the `error.cwl` file?

**Comment by ALuesink** [comment]
> @ALuesink do you still have the `error.cwl` file?

Yes, I have it. However, I now see it has no tabs so it would not create the error. I've tried a file with tabs instead of spaces, but I don't get the `found character \t` error. This time I get `'NoneType' object has no attribute ...`. 
Is it possible the error has changed in the meantime?

**Comment by gcapes** [comment]
Thanks for checking. I get the same error I when trying to reproduce an example with tabs. I'll add an example to this episode next week.


Thanks
Gerard
--
Gerard Capes
Research Applications, IT Services, University Of Manchester
Hours: Tues - Fri, 08:00 - 16:00

________________________________

File: _episodes/debug.md [hunk 5]
```diff
@@ -10,16 +10,389 @@ objectives:
...
+
+Some very common YAML errors are:
+
+- Using tabs instead of spaces. In YAML files indentations are made using spaces, not tabs.
+  Errors caused by tabs will show `'NoneType' object has no attribute 'name'`.
+
+  ~~~
+  cwlVersion: v1.2
+  class: Workflow
+
```
Comment (mr-c) [inline _episodes/debug.md:50]: Maybe a shorter example? This is currently longer than a screenful with not much advantage for that

**Review by mr-c — CHANGES_REQUESTED** [review]

**Comment by gcapes** [comment]
This is now ready for review/merge. The 'wiring' error I haven't done, because I can't actually run the workflow in this lesson. It crashes my laptop part way through.

I would think these would work well as exercises rather than examples, but have followed the pattern in the original submission for now.

File: _episodes/debug.md (hunk 5 above)
Comment (gcapes) [reply _episodes/debug.md:50]: There's a tradeoff between using the example from the rest of the lesson that learners will be familiar with vs using a 'hello world' example which would be shorter but unrelated. What do you think?

**Review by mr-c — APPROVED** [review]

**Pull request merged**